### ISSUE TITLE
Return DocumentFragment when called w/o a tag?

### ISSUE THREAD

**Issue by tomekwi** [dominictarr/hyperscript#37]
I use `DocumentFragment`s quite often when constructing DOM programmatically. And compare these three options – first two are real code, the third is what I propose:

```js
//» `<div>` soup.
h('div', [
  h('.abc'),
  h('.def'),
]);

//» Dreadful DOM spaghetti code. Bleeh…
const fragment = document.createDocumentFragment();
fragment.appendChild(h('.abc'));
fragment.appendChild(h('.def'));

//» How refreshing!
h([
  h('.abc'),
  h('.def'),
]);
```

**Comment by dominictarr** [comment]
I've never used document fragments. what are it's advantages?

**Comment by tomekwi** [comment]
It’s a container for multiple DOM elements.

Say we’re building a dynamic list of things. A container is in the DOM but we need to add a hundred items dynamically. We can create the items quickly outside the document, pack them into a DocumentFragment and attach them to the DOM when all is ready. We touch the document DOM only once instead of touching it a hundred times. That’s what I’ve been using DFs for.

Another example (my current use case). I’m creating *[tape-css](https://github.com/studio-b12/tape-css)* for fast CSS unit tests. The API takes an Element when you want one thing attached for a test – or a DocumentFragment when you want more. I recommend *hyperscript* there to make the tests more readable – it’d be good if it could cover both options.

**Comment by dominictarr** [comment]
@tomekwi would it be the same as adding them as children of a new element, then adding that one element?

It would be more consistent to support `h(el1, el2, el3) => DocumentFragment` this means that hyperscript would also need to check wether an element it has been passed is a document fragment too.

I've also been considering `h(document.body, ...)` to be the element that is then updated (appended to) as shortcut for `document.body.appendChild(h(...))` but maybe that is bad because it's mutating.

The other question, is there a sensible character to do something like `h('#', ...)` to create a DocumentFragment? I am awaiting a pr that allows you to create a comment with `h('!', ...)`

**Comment by tomekwi** [comment]
> would it be the same as adding them as children of a new element, then adding that one element?

Yeah from an API point of view that’s exactly the same thing – except the intermediate element doesn’t trash your DOM at the end.

```js
  h(selector, options?, (...DOM|Array<DOM>)?) => Element
| h(options?, (...DOM|Array<DOM>)) => DocumentFragment

DOM: Element|Text|DocumentFragment
```

(That’s [jsig](http://jsig.biz).) I guess that’s a sensible option. The DocumentFragment is no real element – it doesn’t get added to the DOM at the end. It’s just a container which gets flushed as soon as you add it as a child: http://jsbin.com/voluwox/1/edit?html,js,console,output. Unlike a comment which is actually a real Node (http://devdocs.io/dom/document/createcomment). So I’d say —leave out the selector to get a DocumentFragment at the end.

I hope I’ll have some time to help you with that. But probably no sooner than in three weeks or so – I have deadlines here and there before I go off on vacation.

**Comment by tomekwi** [comment]
> but maybe that is bad because it's mutating

Using, designing and documenting APIs I learned that it’s good when one function returns a value *or* mutates things. But not both at the same time.

**Comment by dominictarr** [comment]
yeah, that is a good point.

**Comment by henriklundgren** [comment]
my pr follows OPs suggestion, i.e.
```js
h([
  h('.fragment')
]);
```

**Comment by tomekwi** [comment]
@henriklundgren thanks for tackling this! :beer:

**Comment by dominictarr** [comment]
sorry I forgot about this. I feel abit uncomfertable about the syntax here. it feels like a special case, and maybe [] as the first argument will be useful for something else? would it be a pain in the but to make it `h.frag(...)` or `h.docfrag(...)` or similar?

**Comment by tomekwi** [comment]
```js
h('div', [h('a'), h('a')]));
//» <div>
//    <a>
//    <a>
//  </div>
```

```js
h(/* nothing, */ [h('a'), h('a')]));
//» <!-- nothing -->
//    <a>
//    <a>
//  <!-- /nothing -->
```

I guess it’s quite logical as it is. But I’m fine either way.

**Comment by dominictarr** [comment]
well, the array is actually optional, so you can do `h('div', h('a'), h('a')));` that means that it's the array that has a specal meaning in this case, which is what makes me uncomfertable.

**Comment by henriklundgren** [comment]
Someone that uses fragments should chime in and review it.

```js
import h from 'hyperscript';

// Maybe
h.createFragment = function(arr) {
  return arr.reduce(function(prev, next) {
    const frag = prev;
    frag.appendChild(next);
    return frag;
  }, document.createDocumentFragment());
};

const second = [1,2,3,4,5,6,7,8,9,10].map(function(_v, idx) {
  return h('li', { className: 'list-name'}, `item-${_v}`);
});

const list = h('ul', {className:'new-list'}, h.createFragment(second));
document.body.appendChild(list);
```

**Comment by tomekwi** [comment]
@henriklundgren does `h(h('a'), h('a'))` work as well?

(BTW, I actually prefer the array syntax for its kinda-fixed-arity explicitness.)

**Comment by tomekwi** [comment]
But I’m fine with `h.createFragment` as well.
